aws/transport/http/response_error_middleware.go

```diff
@@ -0,0 +1,63 @@
+package http
+
+import (
+	"context"
+
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+)
+
+// AddResponseErrorWrapper adds response error wrapper middleware
+func AddResponseErrorWrapper(stack *middleware.Stack) {
+	// add error wrapper middleware before operation deserializers so that it can wrap the error response
+	// returned by operation deserializers
+	stack.Deserialize.Insert(&errorWrapperMiddleware{}, "OperationDeserializer", middleware.Before)
+}
+
+type errorWrapperMiddleware struct {
+}
+
+// ID returns the middleware identifier
+func (m *errorWrapperMiddleware) ID() string {
+	return "AWSResponseErrorWrapperMiddleware"
+}
+
+func (m *errorWrapperMiddleware) HandleDeserialize(ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler) (
+	out middleware.DeserializeOutput, metadata middleware.Metadata, err error,
+) {
+	out, metadata, err = next.HandleDeserialize(ctx, in)
+	if err == nil {
+		// Nothing to do when there is no error.
+		return out, metadata, err
+	}
+
+	resp, ok := out.RawResponse.(*smithyhttp.Response)
+	if !ok {
+		// No raw response to wrap with.
+		return out, metadata, err
+	}
+
+	// look for request id in metadata
+	reqID := GetRequestIDMetadata(metadata)
+	if len(reqID) == 0 {
+		// check for request id in header `X-Amzn-Requestid`
+		if v := resp.Header.Get("X-Amzn-Requestid"); len(v) != 0 {
+			reqID = v
+			// set reqID on metadata for successful responses.
+			SetRequestIDMetadata(&metadata, reqID)
+		}
+	}
+
```

Comment:
There is also the `X-Amz-RequestId` header as well i think.

Also this middleware probably shouldn't do double duty and check for the header it self. It should rely on the metadata to have a value for the RequestID. Duplicating this behavior makes it difficult to reason about where things are coming from.